### ISSUE TITLE
Python: Client connecting to gRPC server takes 20 seconds (IPv6 fallback to IPv4)

### ISSUE THREAD

**Issue by ben-xD** [grpc/grpc#31466]
When connecting using a network that doesn't support IPv6, (but both client and server support IPv6, and a `AAAA` record exists for that server), I've seen a 20 second delay in connecting. I believe gRPC is attempting IPv6 but fails, and then backs off for **19 seconds**, and then fails again, then tries IPv4, which then works immediately. ~~Sometimes, it doesn't connect at all.~~

When using the IPv4 address directly, there is **no delay**.

**Workaround:** If I could disable IPv6, I can workaround this, but I don't see an option for gRPC to do this. I'd have to just convert the domain name to an IPv4 record manually, and use that. 

**TLDR:** It takes 20 seconds for the Python client to connect to the gRPC server when IPv6.

### Configuration
- Set environment variables:
```
GRPC_VERBOSITY=DEBUG
```

### What version of gRPC and what language are you using?
- `grpc-python/1.50.0`
- Initially tested on `grpc-python/1.48.0`, also with the same issue, but slightly different log lines

### What operating system (Linux, Windows,...) and version?
macOS Monterey 12.6

### What runtime / compiler are you using (e.g. python version or version of gcc)
Tested on both Python 3.9.11 and 3.10.6

### What did you expect to see?
Fallback to IPv4 faster (immediately)

### What did you see instead?

Relevant logs: specifically, `connect failed (UNKNOWN:tcp handshaker shutdown {created_time:"2022-10-25T16:46:35.080532+01:00"}), backing off for -19002 ms`
```
D1026 13:08:21.277694000 4316972416 config.cc:113]                     gRPC EXPERIMENT tcp_frame_size_tuning               OFF (default:OFF)
D1026 13:08:21.277777000 4316972416 config.cc:113]                     gRPC EXPERIMENT tcp_read_chunks                     OFF (default:OFF)
D1026 13:08:21.277780000 4316972416 config.cc:113]                     gRPC EXPERIMENT tcp_rcv_lowat                       OFF (default:OFF)
D1026 13:08:21.277782000 4316972416 config.cc:113]                     gRPC EXPERIMENT peer_state_based_framing            OFF (default:OFF)
D1026 13:08:21.277785000 4316972416 config.cc:113]                     gRPC EXPERIMENT flow_control_fixes                  OFF (default:OFF)
D1026 13:08:21.277787000 4316972416 config.cc:113]                     gRPC EXPERIMENT memory_pressure_controller          OFF (default:OFF)
D1026 13:08:21.277788000 4316972416 config.cc:113]                     gRPC EXPERIMENT periodic_resource_quota_reclamation OFF (default:OFF)
D1026 13:08:21.277790000 4316972416 config.cc:113]                     gRPC EXPERIMENT unconstrained_max_quota_buffer_size OFF (default:OFF)
D1026 13:08:21.277792000 4316972416 config.cc:113]                     gRPC EXPERIMENT new_hpack_huffman_decoder           OFF (default:OFF)
D1026 13:08:21.277793000 4316972416 config.cc:113]                     gRPC EXPERIMENT event_engine_client                 OFF (default:OFF)
D1026 13:08:21.277921000 4316972416 ev_posix.cc:141]                   Using polling engine: poll
D1026 13:08:21.277927000 4316972416 dns_resolver_ares.cc:831]          Using ares dns resolver
D1026 13:08:21.278006000 4316972416 lb_policy_registry.cc:45]          registering LB policy factory for "priority_experimental"
D1026 13:08:21.278011000 4316972416 lb_policy_registry.cc:45]          registering LB policy factory for "weighted_target_experimental"
D1026 13:08:21.278013000 4316972416 lb_policy_registry.cc:45]          registering LB policy factory for "pick_first"
D1026 13:08:21.278015000 4316972416 lb_policy_registry.cc:45]          registering LB policy factory for "round_robin"
D1026 13:08:21.278018000 4316972416 lb_policy_registry.cc:45]          registering LB policy factory for "ring_hash_experimental"
D1026 13:08:21.278064000 4316972416 lb_policy_registry.cc:45]          registering LB policy factory for "grpclb"
D1026 13:08:21.278116000 4316972416 lb_policy_registry.cc:45]          registering LB policy factory for "rls_experimental"
D1026 13:08:21.278136000 4316972416 lb_policy_registry.cc:45]          registering LB policy factory for "xds_cluster_manager_experimental"
D1026 13:08:21.278140000 4316972416 lb_policy_registry.cc:45]          registering LB policy factory for "xds_cluster_impl_experimental"
D1026 13:08:21.278143000 4316972416 lb_policy_registry.cc:45]          registering LB policy factory for "cds_experimental"
D1026 13:08:21.278145000 4316972416 lb_policy_registry.cc:45]          registering LB policy factory for "xds_cluster_resolver_experimental"
D1026 13:08:21.278148000 4316972416 certificate_provider_registry.cc:35] registering certificate provider factory for "file_watcher"
I1026 13:08:42.996989000 6226325504 subchannel.cc:910]                 subchannel 0x142fe0610 {address=ipv6:%IPV6_ADDRESS:8000, args={grpc.client_channel_factory=0x122f2bc00, grpc.default_authority=example.com:8000, grpc.internal.channel_credentials=0x122f2da00, grpc.internal.security_connector=0x142fc7000, grpc.internal.subchannel_pool=0x122f805d0, grpc.primary_user_agent=grpc-python/1.50.0, grpc.resource_quota=0x122f2e2c0, grpc.server_uri=dns:///example.com:8000}}: connect failed (UNKNOWN:tcp handshaker shutdown {created_time:"2022-10-26T13:08:41.989482+01:00"}), backing off for -19003 ms
I1026 13:08:42.998464000 6227472384 subchannel.cc:867]                 subchannel 0x142fe0610 {address=ipv6:%IPV6_ADDRESS:8000, args={grpc.client_channel_factory=0x122f2bc00, grpc.default_authority=example.com:8000, grpc.internal.channel_credentials=0x122f2da00, grpc.internal.security_connector=0x142fc7000, grpc.internal.subchannel_pool=0x122f805d0, grpc.primary_user_agent=grpc-python/1.50.0, grpc.resource_quota=0x122f2e2c0, grpc.server_uri=dns:///example.com:8000}}: backoff delay elapsed, reporting IDLE
```

### Anything else we should know about your project / environment?
The applications are connected by an OpenVPN network.

**Comment by gnossen** [comment]
@yashykt Any insight into this behavior?

**Comment by yashykt** [comment]
Without more verbose logging, my suspicion is that the initial connection attempt itself is taking 20 seconds.

**Comment by apolcyn** [comment]
If both IPv4 and IPv6 are resolved, and the IPv6 addresses are locally routable, then the client will prefer IPv6  because grpc's default DNS resolver follows RFC 6724.

Ideally, if the IPv6 addresses are not reachable, then the local kernel shouldn't have route table entries for them, otherwise they'll always act as black holes.

If that's not changeable, then another knob could be to set the system-wide [gai.conf](https://man7.org/linux/man-pages/man5/gai.conf.5.html) file, which can be used to override the address preferences used by `getaddrinfo`. gRPC does not use `getaddrinfo` by default, but it will use it if you set the env variable `GRPC_DNS_RESOLVER=native`.

Another clunky workaround could be to set a very low connection timeout in gRPC, to reduce the time wasted on those IPv6 addresses (almost along the lines of happy eyeballs).

**Comment by markdroth** [comment]
In general, the subchannel connection backoff delays (the ones that you're seeing the log messages for) should not be relevant here, because those delays kick in only after the connection attempt has failed.  As soon as the connection attempt fails, the pick_first LB policy (the default) moves on to the next address in the list and tries to connect to that.  So by the time you see that log message, it's no longer really relevant.

Note that the backoff delay can be negative if the connection attempt takes longer than the computed time of the next attempt, which is computed when the previous attempt is started.  The [backoff algorithm](https://github.com/grpc/grpc/blob/master/doc/connection-backoff.md) has a min connect timeout that trumps the backoff delay computation.  In this case, the timer will fire immediately, meaning that we do not impose any delay in attempting to reconnect to that particular address.

So I suspect the problem here is actually the min connect time (20 seconds by default), not the backoff time, which is what the log message is about.  I believe that the -19 second backoff delay being logged here is a reflection of the fact that the connection attempt for the initial IPv6 address took 20 seconds, whereas the computed start time for the next attempt would have been around a second.

I think @apolcyn is correct that if the IPv6 address is not routable, then it should try the IPv4 address first.  So the core problem here is probably that your system has routes for the IPv6 addresses even though it shouldn't, and I think that's your best bet for fixing this.

If that won't work for some reason, then the other work-arounds that Alex suggested are your next best bet.

I will also note that we are planning to implement happy eyeballs support in Q1 next year, which should reduce the timeouts in this situation.

I hope this info is helpful.

**Comment by ben-xD** [comment]
Thanks for your responses. It's interesting to read (and read again 😅). Looks like it will come in handy in the future.

I agree, in our case, the network (OpenVPN) sets `AAAA` record for services running on devices which don't support IPv6. It looks like there isn't really a way to disable these `AAAA` records. Another temporary workaround is to disable IPv6 on client machines.

Instead of adjusting the python clients, we are solving the issue by avoiding OpenVPN: there shouldn't be AAAA records if IPv6 isn't supported.

**Comment by ben-xD** [comment]
Sorry, I should've said 🤓: there shouldn't be AAAA records for the service if IPv6 isn't supported by the service.

I only face this issue for developer machines, which use IPv6, requesting a service that doesn't support IPv6. All the production machines have IPv6 disabled. 

I can and have done `a)` for my personal machine (OpenVPN allows you to set "IPv4 only" in it's advanced settings). However, I don't have much control over the network of other developers' machines and wanted to solve this for everyone. 

Presumable `dig AAAA storage.googleapis.com` works because `storage.googleapis.com` supports IPv6. However, in my case `dig AAAA myservice.com` resolves to an IPv6 address that will never work because the service doesn't actually support IPv6.

I'm taking a look at that RFC and other options (blocking that IPv6 address locally).
